### PR TITLE
Homework 8 pull request

### PR DESCRIPTION
**Pull request by ilkivs**

### CODE REVIEW

**Comment by ilkivs** [comment]
Wrong homework number
It is homework #9

File: src/HomeWork/hw8/Mail/MailMessage.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package HomeWork.hw8.Mail;
+
+public class MailMessage implements Sendable<String> {
+
+    private String from;
```
Comment (Siril-O) [inline src/HomeWork/hw8/Mail/MailMessage.java]: can be final

File: src/HomeWork/hw8/Mail/MailService.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+public class MailService<T> implements Consumer<Sendable<T>> {
+
+    private Map<String, List<T>> mailBox;
+
+    public MailService() {
+        mailBox = new HashMap<>() {
+            @Override
+            public List<T> get(Object key) {
+                return super.get(key) == null ? new ArrayList<>() : super.get(key);
```
Comment (Siril-O) [inline src/HomeWork/hw8/Mail/MailService.java]: better move this logic where it is use without overriding HashMap

File: src/HomeWork/hw8/Mail/Salary.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package HomeWork.hw8.Mail;
+
+public class Salary implements Sendable<Integer> {
+
+    private String from;
```
Comment (Siril-O) [inline src/HomeWork/hw8/Mail/Salary.java]: make final you do not have setters

File: src/HomeWork/hw8/Mail/Test.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
+package HomeWork.hw8.Mail;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class Test {
+
+    public void test() {
```
Comment (Siril-O) [inline src/HomeWork/hw8/Mail/Test.java]: create separate method for each test case

File: src/HomeWork/hw8/MinMaxValue.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class MinMaxValue {
+
+    public static <T> void findMinMax(Stream<? extends T> stream, Comparator<? super T> order,
+                                      BiConsumer<? super T, ? super T> minMaxConsumer) {
+
+        List<T> streamList = stream.sorted(order).collect(Collectors.toList());
```
Comment (Siril-O) [inline src/HomeWork/hw8/MinMaxValue.java]: Why do you sort before finding min/max?

File: src/HomeWork/hw8/NumberGenerator.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    https://stepik.org/lesson/12781/step/11?unit=3128
+ */
+
+public class NumberGenerator {
+
+    private static int currentNumber;
+    private static int index = 0;
+
+    public static IntStream pseudoRandomStream(int seed) {
+        currentNumber = seed;
```
Comment (Siril-O) [inline src/HomeWork/hw8/NumberGenerator.java]: check seed

File: src/HomeWork/hw8/NumberGenerator.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        return IntStream.generate(() -> mid(currentNumber, index++));
+    }
+
+    private static int mid(int number, int index) {
+        if (index == 0) {
+            return number;
+        }
+        number *= number;
```
Comment (Siril-O) [inline src/HomeWork/hw8/NumberGenerator.java]: do not change input argument, this is very bad practice

File: src/HomeWork/hw8/NumberGenerator.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        int first = 0, second = 0, third = 0;
+        if (number > 9) {
```
Comment (Siril-O) [inline src/HomeWork/hw8/NumberGenerator.java]: make generic method for this 3 cases

**Review by Siril-O — CHANGES_REQUESTED** [review]

File: src/HomeWork/hw8/Mail/MailService.java (hunk 2 above)
Comment (ilkivs) [reply src/HomeWork/hw8/Mail/MailService.java]: i changed this, but now i have 1 failed assert cause NPE (
class Test => method MessageStreamTest() => commented line
mailBox.get(randomTo) = null and i dont know how to solve it correctly

**Pull request merged**